Re-organize variant edit form and flow

mamhoff opened the pull request:
In order to add somewhere to edit prices, I need the variant form
to be somewhat more structured than it has been in the past. Using
nested `fieldset` elements with `legend`s, this commit breaks the
edit variant form up in a "general", a "pricing" and a "shipping"
section, each with the relevant input fields.

In order to be consistent with the search form on the index page,
I chose a four-column layout.

All `data-hook`s are still there, although I moved some of them.

I also improved I18n, using the t('.something') helper which looks
translations up according to the name of the partial from which the
`t` is called. This pattern has been (sort of) used in the user and store
credit partials already, and does really good namespacing.

**New search box, and `deleted` checkbox next to it**
![solidus administration variants 2016-04-05 18-08-25](https://cloud.githubusercontent.com/assets/703401/14289200/1082669c-fb5a-11e5-90bc-7496d14e8186.png)

**`New Variant` form, now full-width**
![solidus administration variants 2016-04-05 18-08-54](https://cloud.githubusercontent.com/assets/703401/14289201/1083fce6-fb5a-11e5-9197-1f8bbe012693.png)

**`Edit Variant` form, now full-width, without allowing the user to edit option values** 
![solidus administration variants 2016-04-05 18-09-19](https://cloud.githubusercontent.com/assets/703401/14289202/10853872-fb5a-11e5-956d-914b4db3531b.png)

**When checking the `Show deleted variants` checkbox, deleted are shown in red**
![solidus administration variants 2016-04-05 18-09-40](https://cloud.githubusercontent.com/assets/703401/14289198/1081c4a8-fb5a-11e5-9eb6-2f0e9466fe9f.png)

**Link to product details page if no option values are set for this product**
![solidus administration variants 2016-04-05 18-10-49](https://cloud.githubusercontent.com/assets/703401/14289199/10823be0-fb5a-11e5-8fb4-995c1d2a7b5c.png)

graygilmore commented:
Nice change. Probably a good candiate for tabs.

mamhoff commented:
I was actually surprised to see that the product form doesn't have them, yet! I do think you mean the sidebar, not the sections in this form, correct?

mamhoff commented:
I talked to @Mandily about this IRL, and I'll be updating this with lots of her input.

mamhoff commented:
Done updating. This has come out pretty nice! I have a few questions for reviewers, which I'll put as comments to individual code lines. Also: Check out the screencasts!

mamhoff commented on backend/app/assets/stylesheets/spree/backend/shared/_forms.scss:
```diff
@@ -139,6 +139,10 @@ span.info {
 
+  &.centered {
+    margin-top: 13px;
+  }
```
I don't like this, but if I don't do something like that the checkbox that it's applied to ("Track inventory") is way too far up. I'm sure there's a better way to do this. Frontenders?

graygilmore replied:
Try adding the `.checkbox` class to the div instead.

mamhoff replied:
Perfect! Thanks loads!

mamhoff commented on backend/app/views/spree/layouts/admin.html.erb:
```diff
@@ -19,7 +19,7 @@
               </div>
             <% end %>
-            <div class="<%= if content_for?(:sidebar) then 'twelve' else 'sixteen' end %> columns">
+            <div id="content-without-nav" class="<%= if content_for?(:sidebar) then 'twelve' else 'sixteen' end %> columns">
```
This is an unfortunate name, but `content` has already been taken by the div above that contains both the `aside` nav and the actual content. If anyone comes up with a better name, I'd be happier.

graygilmore replied:
Totally. I have completely reworked this for future layouts because I ran into the same problem: https://github.com/solidusio/solidus/compare/master...graygilmore:content_layout#diff-b1a47ee5eb885f91f6fe5ff9b2be5211R13

Since `#content-without-nav` is only being used for a JS hook maybe just something like `class="js-content-below-tabs`? Either way it's going to be ugly but I'd rather avoid something like `content-without-nav` as it implies that there is no nav on the page (if you were viewing it out of context, like in the `new.js.erb` file).

mamhoff commented on core/config/locales/en.yml:
```diff
@@ -715,6 +716,18 @@ en:
+          dimensions: Dimensions
+          no_tax_category_override: No override
+          pricing: Pricing
+          pricing_hint: These values are populated from the product details page and can be overridden below
```
This is what happens if you use the t('.something') convention. It's nice, and very well namespaced. We're losing a bit of generality here, but on the other hand it's very easy to find where to add a translation. Loved it.

tvdeyen replied:
That's nice!

mamhoff commented:
@graygilmore @Mandily Do have a look, I rebased with the current master - now we have: 

- Full-width table 
- Full-width forms

This is from older PRs:
- Working search in variants
- Tabs :confetti_ball: 
- Consistent view of deleted variants
- Deleted variants don't lose their values anymore

See updated screen shots in the description

tvdeyen commented:
@mamhoff the screenshots above are also updated?

mamhoff commented:
@tvdeyen Yes they should be, does it look old to you?

tvdeyen commented:
No. Was just asking, because you made changes to the layout according to your last comment.  Thanks

tvdeyen commented on backend/app/controllers/spree/admin/variants_controller.rb:
```diff
@@ -31,12 +31,10 @@ def new_before
       end
 
       def collection
-        @deleted = (params.key?(:deleted) && params[:deleted] == "on") ? "checked" : ""
```
Do people depend on that instance var?

mamhoff replied:
I don't think so. It was only used for filtering, and that's been entirely revamped (we already "lost" `@option_types` and `@option_values`).

Mandily commented:
It seems to me that "EDIT VARIANT - SIZE: M, COLOR: BLUE" should actually be the title of this screen, instead of the title of the first section. Is there any chance we can add that to the page title? I'm not sure if we need the variant type in the title though, what do you guys think?
(Eventually I'd like to get some breadcrumb functionality in the title instead of having the 'back to products list' button. But that's probably a little bit of scope creep for this thread)

<img width="503" alt="short title" src="https://cloud.githubusercontent.com/assets/15271353/14292850/bd20bf2a-fb1e-11e5-8e41-1554024368f0.png">
<img width="618" alt="long title" src="https://cloud.githubusercontent.com/assets/15271353/14292851/c00f33e2-fb1e-11e5-8326-9941a2d75881.png">

mamhoff commented:
Yes, @Mandily that's possible. Look!

![solidus administration variants 2016-04-05 20-58-49](https://cloud.githubusercontent.com/assets/703401/14294420/4cf22bfa-fb71-11e5-89cf-61afc0468444.png)

tvdeyen commented on backend/app/views/spree/admin/shared/_product_tabs.html.erb:
```diff
@@ -1,5 +1,5 @@
 <% content_for :page_title do %>
-  <%= Spree.t(:editing_product) %> &ldquo;<%= @product.name %>&rdquo;
+  <%= Spree::Product.model_name.human(count: :other) %> &#x2f; <%= @product.name %>
```
:smirk:

mamhoff commented:
Yeah, let's do the breadcrumbing in a different PR. :)

Mandily commented:
I really like the page title update, that's harder to miss than where it was before. I have 3 more ideas for improving this:
1. I know I had section titles in my wires, but I think it's cleaner, takes up less space and still gets the job done without them. 
2. It would be more consistent with the rest of the app if we could label the 'override tax category" as 'variant tax category', and then have the default option be 'default'
3. Shorten the width of the tax category field

![cleaned up](https://cloud.githubusercontent.com/assets/15271353/14297911/50906fdc-fb36-11e5-90fb-a5858c006f1c.jpg)

mamhoff commented:
1. Can we leave the "Edit Variant" in? I like the frame there, aesthetically :) Otherwise I'm fine with killing the sections headers and additional lines. 

2. I don't like calling anything "Default", because default is never actually defined and can mean different things in one or the other context. I could label the "No override option" with the tax category name from the product, but that's not really what we're storing in the database (we're storing nothing in this case). I'm curious for other opinions here. 

3. NP, will do.

mamhoff commented:
@Mandily here's the form without `legend`s - Still much better than what we had before! There's a larger gap towards the second row because the checkbox needs more space around itself than the other fields, but that's not something I want to tackle here. 

I went with "Use Product Tax Category" instead of "No override", which I think is understandable.

![solidus administration variants 2016-04-07 00-49-32](https://cloud.githubusercontent.com/assets/703401/14335374/efeb0a5e-fc5a-11e5-9f4e-44f1b4e82c37.png)

Mandily commented:
This looks great! The "Use Product Tax Category" text is much clearer and it looks closer to fitting on one screen.  :+1:

jhawthorn commented:
:+1:
